src/dsa/rpo_falcon512/keys/secret_key.rs

```diff
@@ -0,0 +1,327 @@
+                .iter()
+                .map(|a| a.re.round() as i16)
+                .collect::<Vec<i16>>()
+                .try_into()
+                .expect("The number of coefficients should be equal to N");
+
+            if let Ok(s1) = SignaturePoly::try_from(&s1_coef) {
+                if let Ok(s2) = SignaturePoly::try_from(&s2_coef) {
+                    if s2.fft().coefficients.iter().all(|&c| c != FalconFelt::zero()) {
+                        break (s1, s2);
+                    }
+                }
+            }
+        };
+        Ok(s)
+    }
+}
+
+// SERIALIZATION / DESERIALIZATION
+// ================================================================================================
+
+impl Serializable for SecretKey {
+    fn write_into<W: ByteWriter>(&self, target: &mut W) {
+        let basis = &self.secret_key;
+
+        // header
+        let n = basis[0].coefficients.len();
+        let l = n.checked_ilog2().unwrap() as u8;
+        let header: u8 = (5 << 4) | l;
+
+        let f = &basis[1];
+        let g = &basis[0];
+        let capital_f = &basis[3];
+
+        let mut buffer = Vec::with_capacity(1281);
+        buffer.push(header);
+
+        let f_i8: Vec<i8> = f.coefficients.iter().map(|&a| -a as i8).collect();
+        let f_i8_encoded = encode_i8(&f_i8, 6).unwrap();
+        buffer.extend_from_slice(&f_i8_encoded);
+
+        let g_i8: Vec<i8> = g.coefficients.iter().map(|&a| a as i8).collect();
+        let g_i8_encoded = encode_i8(&g_i8, 6).unwrap();
+        buffer.extend_from_slice(&g_i8_encoded);
+
+        let big_f_i8: Vec<i8> = capital_f.coefficients.iter().map(|&a| -a as i8).collect();
+        let big_f_i8_encoded = encode_i8(&big_f_i8, 8).unwrap();
+        buffer.extend_from_slice(&big_f_i8_encoded);
+        target.write_bytes(&buffer);
+    }
+}
+
+impl Deserializable for SecretKey {
+    fn read_from<R: ByteReader>(source: &mut R) -> Result<Self, DeserializationError> {
+        let byte_vector: [u8; SK_LEN] = source.read_array()?;
+
+        // check length
+        if byte_vector.len() < 2 {
+            return  Err(DeserializationError::InvalidValue("Invalid encoding length: Failed to decode as length is different from the one expected".to_string()));
+        }
+
+        // read fields
+        let header = byte_vector[0];
+
+        // check fixed bits in header
+        if (header >> 4) != 5 {
+            return Err(DeserializationError::InvalidValue("Invalid header format".to_string()));
+        }
+
+        // check log n
+        let logn = (header & 15) as usize;
+        let n = 1 << logn;
+
+        // match against const variant generic parameter
+        if n != N {
+            return Err(DeserializationError::InvalidValue(
+                "Unsupported Falcon DSA variant".to_string(),
+            ));
+        }
+
+        let width_f = field_element_width(0);
+        let width_g = field_element_width(1);
+        let width_big_f = field_element_width(2);
```

Comment:
Should these be constants? Seems like `field_element_width` just returns either 6 or 8 depending on the passed in value.

Reply:
Fixed